회원가입 및 로그인 관련 중요 코드 수정, Jwt에러코드 관련 수정

djlim00 opened the pull request:
.## 📚 개요

회원가입 및 로그인 관련 중요 코드 수정, Jwt에러코드 관련 수정을 했습니다.
아래에 수정 내용 적어두었습니다!

## 🕐 리뷰 예상 시간

> 30m

## ❗❗ 중요한 변경점(Option)

- 회원 탈퇴 기능을 구현하기 앞서서 회원의 게시글, 좋아요, 댓글을 전부 삭제하기 위해서 연관관계 매핑을 수정했습니다.

- 회원가입 및 로그인 기능을 기존에 하나에 묶여있던걸 따로 나누었습니다.
- 회원가입이 완료된 이후에 온보딩을 통해서 아르바이트 경력, 관심 아르바이트(추후 구현)를 회원 정보에 입력하도록 수정했습니다.
- 닉네임 중복 여부를 확인하는 api를 추가했습니다.

- JWT 관련 에러 목록을 추가했습니다.
- 회원가입과 로그인을 할 때는 Jwt필터에 걸리지 않도록 수정했습니다.

코드관련해서 질문 사항 있으시면 언제든지 코멘트 달아주세요~ 디코도 가능합니다!

djlim00 committed:
update: 게시글, 좋아요, 댓글 연관관계 매핑 수정

- User와 연관관계를 맻도록 수정했습니다.

djlim00 committed:
feat: 회원탈퇴 기능 구현

- 회원이 탈퇴하게 되면, 회원이 생성한 게시글, 댓글, 좋아요, 게시글 이미지가 삭제되도록 구현했습니다.

djlim00 committed:
update: 회원가입 및 로그인 기능 수정

- 온보딩으로 가져오는 경력 및 관심 목록(카테고리 구현완료 후 구현예정) 가져오는 api 따로 구현
- 기존의 모로한 로그인과 회원가입 api 분리

djlim00 committed:
Merge remote-tracking branch 'origin/main' into feature/-kakao-app-login

djlim00 committed:
feature: 닉네임 중복확인 api 구현

- 회원가입 시 닉네임 중복을 위한 api 구현했습니다.

djlim00 committed:
update: 닉네임 중복 반환값 수정

- isDuplicate를 boolean 값으로만 반환하는 것 보다 명확하게 반환할 수 있도록 수정했습니다.

djlim00 committed:
update: Jwt관련 에러 코드 추가

- 기존의 false만 반환하는 방식에서
-  ACCESS_TOKEN_EXPIRED 추가

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -31,7 +31,7 @@ class ArticleService(
                 likeCount = 0,
                 isAnonymous = request.isAnonymous,
-                userId = request.userId,
+                user = userService.getUser(request.userId),
```
user를 위에서 변수로 할당하고 여기서 그 변수를 할당해주시죠

djlim00 replied:
'
	val user = userService.getUser(request.userId)
	
	val article = articleRepository.save(
            Article(
                title = request.title,
                text = request.text,
                likeCount = 0,
                isAnonymous = request.isAnonymous,
                user = user,
                articleKindId = request.articleKind,
            )
        )'
 이렇게 하라는 말씀이시죠?

Ji-InPark replied:
넵 맞습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/article/service/ArticleService.kt:
```diff
@@ -82,7 +82,7 @@ class ArticleService(
             .orElseThrow { CustomException(CustomError.ARTICLE_NOT_FOUND) }
 
-        val user = userService.getUser(article.userId)
+        val user = userService.getUser(article.user.id)
```
```suggestion
        val user = article.user
```

이렇게 하지 않고 db에 한 번 더 거치는 이유가 있나요??

djlim00 replied:
아 원래는 article에 userId:Long로 가지고 있던걸 이번에 User테이블이랑 연관관계를 만들면서 연관되는 코드들 수정하는 과정에서 실수한 것 같습니다. 감사합니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/KakaoUserResponse.kt:
```diff
@@ -1,25 +1,55 @@
+    @JsonProperty("kakao_account")
+    val kakaoAccount: KakaoAccount
 ) {
+    @JsonIgnoreProperties(ignoreUnknown = true)
```
https://siahn95.tistory.com/173

필요없다고 하네요

djlim00 replied:
갑자기 카카오 유저 정보가 잘 안받아져서 썼는데 별 의미 없는 것 같더라니.. 감사합니다! 결론적으로는 파라미터명을 공식문서랑 다르게 써서 안받아오던 거였네요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/OnboardingRequest.kt:
```diff
@@ -0,0 +1,10 @@
+    val experienceYears: Int,
+    val experienceMonths: Int,
+
+    // TODO - 카테고리 생성 후 구현
+    // val jobInterests: List<String>? = null
```
main에 카테고리 enum은 머지되어있습니다. 그거 가져다가 쓰시면 됩니다.

djlim00 replied:
구현했습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/KakaoAuthController.kt:
```diff
@@ -19,40 +17,68 @@ class KakaoAuthController(
-        val kakaoAccessToken = kakaoAuthService.getKakaoAccessToken(code)
+    @Operation(
+        summary = "카카오 로그인과 기존 유저 판별",
+        description = "카카오 AccessToken을 헤더로 받아와 로그인하거나, 회원이 아니면 409와 카카오유저정보를 반환합니다." +
+                "반환받은 카카오 유저 정보를 통해서 회원가입에 사용하시면 됩니다."
```
로그인하거나 카카오유저정보를 반환한다는게 이상하게 들리네요.
설명이 명확하지 않은 것 같습니다. 그리고 409를 반환하는 이유가 있나요??

djlim00 replied:
@Operation(summary = "회원가입", description = "회원 가입이 완료되는 시점에 카카오 정보 기반으로 사용자를 DB에 저장합니다.")
으로 수정했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/KakaoAuthController.kt:
```diff
@@ -19,40 +17,68 @@ class KakaoAuthController(
+                "반환받은 카카오 유저 정보를 통해서 회원가입에 사용하시면 됩니다."
+    )
+    @PostMapping("/login")
+    @SwaggerCustomException(CustomExceptionGroup.LOGIN_OR_REGISTER)
```
CustomExceptionGroup도 이름이 변경되어야 할 것 같아요

djlim00 replied:
넵 수정했습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/KakaoAuthController.kt:
```diff
@@ -19,40 +17,68 @@ class KakaoAuthController(
-        val accessToken = jwtProvider.generateAccessToken(user.id.toString())
-        val refreshToken = jwtProvider.generateRefreshToken(user.id.toString())
+        return if (user != null) {
+            val accessToken = jwtProvider.generateAccessToken(user.id.toString())
+            val refreshToken = jwtProvider.generateRefreshToken(user.id.toString())
 
-        return TokenWithUserResponse(
-            accessToken,
-            refreshToken,
-            user
-        )
+            ResponseEntity.ok(TokenWithUserResponse(accessToken, refreshToken, user))
+        } else {
+            ResponseEntity.status(409).body(kakaoUser)
+        }
```
service 해야할 로직들이 controller에 너무나 많네요
서비스 단으로 이동 부탁드립니다.

또한 저렇게 user가 null인 경우에는 CustomException을 throw하는게 더 나을거 같아요.
만약 카카오 유저 정보가 필요한 상황이라면 다른 엔드포인트를 활용하게 하는 것도 좋을 것 같아요.
아니면 다른 엔드포인트가 회원인지 여부를 확인하게끔 하는 것도 좋구요.

djlim00 replied:
반영해서 로직 변경했습니다. 확인부탁드립니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/KakaoAuthController.kt:
```diff
@@ -19,40 +17,68 @@ class KakaoAuthController(
+    fun checkNickNameDuplicate(
+        @RequestParam nickname: String
+    ): NickNameDuplicateResponse {
+        val isDuplicate = kakaoAuthService.isNickNameDuplicate(nickname);
+        val message: String
+
+        if (isDuplicate) message = "닉네임이 중복됩니다."
+        else message = "닉네임이 중복되지 않습니다."
```
service에 있어야할 부분들입니다. 
controller에는 최대한 로직을 넣지 말아주세요

djlim00 replied:
넵 함수 하나만 남기고 전부 Service단으로 넣었습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/KakaoAuthController.kt:
```diff
@@ -19,40 +17,68 @@ class KakaoAuthController(
+    ): OnboardingResponse {
+        val user = kakaoAuthService.updateOnboarding(request)
+        return OnboardingResponse(
+            message = "온보딩 정보가 저장되었습니다.",
```
항상 같은 값을 내려주는거라면 오히려 http status code만으로 충분하다고 생각합니다.

djlim00 replied:
뭔가 저 코드를 작성할 때는 받는 사람 입장에서 명확하게 하고 싶어서 작성했는데,
불필요한 것 같네요 dto에서 userId만 반환하도록 변경했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/KakaoAuthController.kt:
기존 callback url은 사라진건가요??
만약 사라진거라면 application.yml 파일에도 반영해주세요

djlim00 replied:
기존의 백엔드에서 카카오 엑세스 토큰을 획득하는 방식에서는 callback url이 필요했는데, 이제 프론트에서 카카오 엑세스 토큰을 서버에 쏴주는 방식으로 변경되면서 필요없게 되었습니다. 
일단 카카오에 등록된 url이 있긴 한데 백엔드 쪽 코드에서 쓰일 일이 없어서 삭제하도록 하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/util/JwtAuthenticationFilter.kt:
```diff
@@ -19,18 +20,39 @@ class JwtAuthenticationFilter(
 
-            SecurityContextHolder.getContext().authentication = authentication
+            filterChain.doFilter(request, response)
+        } catch (e: CustomException) {
+            response.status = e.customError.status.value()
+            response.contentType = "application/json;charset=UTF-8"
+            val errorJson = """
+            {
+              "code": "${e.customError.name}",
+              "message": "${e.customError.message}"
+            }
+        """.trimIndent()
+            response.writer.write(errorJson)
+            response.writer.flush()
```
꼭 try catch를 써야하는 걸까요??

djlim00 replied:
여기서 try-catch를 할 필요가 없네요.. isValid에서 이미 예외가 나오면 Exception handler로 넘어가니까용
하 예외처리가 자꾸 헷갈리네요 리뷰해주셔서 감사합니다

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,24 @@
+    private val userService: UserService
+) {
+
+    @PostMapping("/withdraw")
```
```suggestion
    @DeleteMapping
```

post를 사용하는 이유가 있나요??

djlim00 replied:
확실히 @DeleteMapping 변경해서 Http 메서드로 구분하는 방법이 훨씬 좋이네용. 기능도 사실상 delete하는 것 외에는 없어서 Post는 별로네요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping("/user")
```
```suggestion
@RequestMapping("/users")
```

djlim00 replied:
인정합니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,24 @@
+    @Operation()
+    fun withdrawUser(
+        @RequestParam userId: Long
+    ): String {
+        userService.deleteUserById(userId)
+        return "회원 탈퇴가 완료되었습니다."
+    }
```
기존대로 response dto사용하게끔 하죠
api가 성공했을 때 항상 message가 같다면 전 http status code로 충분하다는 생각입니다.

djlim00 replied:
response dto를 사용해서 삭제된 회원의 id랑 메세지를 반환하는 것은 어떨까요?

Ji-InPark replied:
그것도 좋습니다만 여전히 저는 message가 달라지지 않는 부분이면 필요한가?에 대해서는 회의적이긴 합니다

djlim00 replied:
인정합니다. 삭제된 userId만 반환하도록 수정할게요.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/entity/User.kt:
```diff
@@ -16,16 +21,16 @@ class User(
-    // @ElementCollection
-    // @Enumerated(EnumType.STRING)
-    // var jobInterests: MutableList<JobCategory> = mutableListOf()
+    @OneToMany(mappedBy = "user", cascade = [CascadeType.ALL], orphanRemoval = true)
+    val articles: MutableList<Article> = mutableListOf(),
+
+    @OneToMany(mappedBy = "user", cascade = [CascadeType.ALL], orphanRemoval = true)
+    val comments: MutableList<Comment> = mutableListOf(),
+    
+    @OneToMany(mappedBy = "user", cascade = [CascadeType.ALL], orphanRemoval = true)
+    val likes: MutableList<ArticleLike> = mutableListOf()
```
마이페이지 때문에 추가하신건가요??

djlim00 replied:
User에서 article,comment,likes랑 연관관계를 맻어놓지 않으면 회원이 탈퇴되었을 때 처리랑, 추후에 마이페이지 만들 때도 필요해서 추가했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/service/UserService.kt:
```diff
@@ -22,4 +22,15 @@ class UserService(
+
+    fun deleteUserById(userId: Long) {
+        val user = userRepository.findById(userId)
+            .orElseThrow { IllegalArgumentException("해당 사용자를 찾을 수 없습니다. id=$userId") }
```
custom exception 사용해주세요

djlim00 replied:
헉 그러게요. 확인해습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/service/UserService.kt:
```diff
@@ -22,4 +22,15 @@ class UserService(
             .orElseThrow { CustomException(CustomError.USER_NOT_FOUND) }
     }
+
+    fun deleteUserById(userId: Long) {
```
delete에는 보통 transacion을 걸어둡니다.

djlim00 replied:
아하 감사합니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/global/exception/CustomException.kt:
val를 안넣었었네요..
어떻게 동작하고 있던거지...

djlim00 replied:
아니 그니까요 저도 처음에 왜 안되지 했었어요.

Ji-InPark reviewed (commented):
한 번 확인해주세요~

djlim00 committed:
update: SwaggerCustomException 수정

djlim00 committed:
Update application.yml

- 카카오 리다리렉트 url이 필요 없어져서 삭제

djlim00 committed:
update: 유저 삭제관련 코드 수정

- deleteUserById에 @Transactional 추가
- Post메서드에서 Delete로 변경
- 유저 삭제시 반환값 dto로 변경

djlim00 committed:
Update KakaoUserResponse.kt

- JsonIgnoreProperties 삭제

djlim00 committed:
update: Retrofit ObjectMapper 변경

djlim00 committed:
update: 회원탈퇴 수정 및 스웨거 명세 추가

djlim00 committed:
Update ArticleService.kt

djlim00 committed:
Update KakaoAuthService.kt

- findUserByKakaoId 예외처리
- 닉네임 중복 로직 Service 단으로 이동

djlim00 committed:
Update CustomExceptionGroup.kt

- 필요없는 에러그룹 삭제

djlim00 committed:
Update JwtAuthenticationFilter.kt

- 필요없는 try-catch문 삭제

djlim00 committed:
update: 엑세스토큰 만료시간 증가

- 원활한 테스트를 위해서 엑세스 토큰 만료 시간을 50분으로 증가

djlim00 committed:
update: 카카오 로그인 및 회원가입 로직 변경

- 카카오 로그인과 관련된 전반적인 로직이 변경되었습니다.
- KaKaoAuthController에서 AuthController로 rename 하면서 다른 클래스명들도 변경되었습니다.
-

djlim00 committed:
chore: getUser() 함수명 변경

- 명확한 의미를 나타내기 위해 함수명을 getUserByUserId()로 변경했습니다.

djlim00 committed:
update: 기존의 Auth도메인 로직 User 도메인으로 이동

- 온보딩 기능과 닉네임 중복확인과 같은 로직들을 User 도메인으로 옮겼습니다.

djlim00 committed:
update: User엔티티 수정

- 바뀐 로직이 기존의 값을 patch 메서드로 수정하기 때문에 val로 선언되었던 값들을 var로 수정했습니다.
- location필드도 String이 아닌 Location객체로 변경 예정입니다.

djlim00 committed:
update: 유저 관심직종 온보딩 저장 기능 구현

- 유저 관심 직종을 온보딩에서 받아오는 로직 구현
- user_job_interest 테이블 추가
- 연관관계 추가

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/AuthService.kt:
```diff
@@ -0,0 +1,84 @@
+@Service
+class AuthService(
+    @Value("\${spring.security.oauth2.client.registration.kakao.client-id}")
+    private val clientId: String,
```
삭제한다고 했었던거 같은데 맞나요??

djlim00 replied:
최신 커밋에는 삭제했는데 반영이 안됐었네요. ym파일에서도 삭제했습니다,

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/AuthService.kt:
```diff
@@ -0,0 +1,84 @@
+    private val kakaoUserApi: KakaoUserApi,
+    private val jwtProvider: JwtProvider
+) {
+    fun getKakaoUserInfo(accessToken: String): KakaoUserResponse {
```
```suggestion
    private fun getKakaoUserInfo(accessToken: String): KakaoUserResponse {
```

밖으로 보여지지 않아도 되는 함수는 private으로 선언해주세요

djlim00 replied:
반영하겠습니다. 감사합니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/AuthService.kt:
```diff
@@ -0,0 +1,84 @@
+        val user = userService.getUserByKakaoId(kakaoUser.id)
+
+        if (user == null) {
+            val user = User(
+                kakaoId = kakaoUser.id,
+                username = kakaoUser.kakaoAccount.name,
+                nickname = kakaoUser.kakaoAccount.kakaoProfile.nickname,
+                profileImageUrl = kakaoUser.kakaoAccount.kakaoProfile.profileImageUrl,
+                ageRange = kakaoUser.kakaoAccount.ageRange,
+                gender = kakaoUser.kakaoAccount.gender,
+                email = kakaoUser.kakaoAccount.email,
+                phoneNumber = kakaoUser.kakaoAccount.phoneNumber
+            )
+            userService.saveUser(user)
+            return RegisterResponse(false, user.id)
```
이렇게만 해도 user에 id가 생기나요??

djlim00 replied:
BaseEntitiy에서 @GeneratedValue(strategy=GenerationType.IDENTITY) 선언해 있어서  User뿐만 아니라 모든 PK가 자동생성되서 들어가요.

Ji-InPark replied:
그건 알고 있는데요 save에서 반환하는 객체가 아니라 기존에 미리 만들어둔 객체에도 적용이 되는건지가 궁금했습니다.

djlim00 replied:
위에서 만들 user도 save()를 하게 되면 JPA에서 `@Id @GeneratedValue(strategy = GenerationType.IDENTITY)`해놓은 필드는 Id 필드를 자동으로 주입을 해줘요. 그래서 아래에서 user.id가 가능합니다.
근데 헷갈릴 수 있어서 그냥 수정할개요~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/AuthController.kt:
```diff
@@ -0,0 +1,55 @@
+        return authService.authorizeWithKakao(kakaoAccessToken)
+    }
+
+    @Operation(summary = "회원가입", description = "회원 가입이 완료되는 시점에 카카오 정보 기반으로 사용자를 DB에 저장합니다.")
+    @PatchMapping("/register")
+    fun register(
+        @RequestBody request: RegisterRequest
+    ): LoginResponse {
+        val user = authService.registerUser(request)
+        val accessToken = authService.generateAccessToken(request.userId)
+        val refreshToken = authService.generateRefreshToken(request.userId)
+
+        return LoginResponse(user.id, accessToken, refreshToken)
+    }
```
지금 구현상으로는 누구든지 userId만 알아내면 다른 회원의 정보를 바꿀 수 있네요.
todo로 jwt 처리 필요하다는 구문만 추가해주세요 머지되고 제가 처리하겠습니다.

djlim00 replied:
네 이거 jwt처리는 나중에 한번에 필요한거 전부 하려고 했는데, 일단 TODO로 적어놓겠습니다.

djlim00 committed:
chore: 카카오엑세스토큰 획득 API 삭제

- 프론트에서 넘겨주는 카카오 엑세스 토큰을 사용하기 때문에 삭제했습니다.

djlim00 committed:
update: 회원 거주지 등록 로직 구현

- 회원의 거주지 등록 로직을 구현했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/AuthorizeResponse.kt:
아래와 같이 한가지 응답으로 통일하죠
기존에 회원이 아니었다고 하더라도 user 객체를 만든 순간부터는 우리의 jwt를 사용하게끔 해야합니다.

```kotlin
	val userId: Long,
	val accessToken: String,
	val refreshToken: String
	val isRegistered: Boolean = false,
```

djlim00 replied:
어제 말씀하신게 이거였군요. 까먹고 있었네용 반영하겠습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/AuthController.kt:
```diff
@@ -0,0 +1,55 @@
+    fun register(
+        @RequestBody request: RegisterRequest
+    ): LoginResponse {
+        val user = authService.registerUser(request)
+        val accessToken = authService.generateAccessToken(request.userId)
+        val refreshToken = authService.generateRefreshToken(request.userId)
+
+        return LoginResponse(user.id, accessToken, refreshToken)
```
```suggestion
        return authService.registerUser(request)
```

controller에서는 최대한 로직을 빼주세요

djlim00 replied:
`RegisterResponse`를 만들어서 `authService.registerUser()`가 `RegisterResponse`를 반환하도록 하겠습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/AuthService.kt:
```diff
@@ -0,0 +1,85 @@
+
+    fun generateRefreshToken(userId: Long): String {
+        return jwtProvider.generateRefreshToken(userId.toString())
+    }
```
이건 없어져도 괜찮을 것 같습니다.

djlim00 replied:
앞선 리뷰에서 Controller에서 로직을 빼고 나니까 저절로 필요없어지네요! 앞으로 Controller 설계할 때 최대한 Service에 모든 로직을 몰아둘 수 있도록 하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/util/JwtAuthenticationFilter.kt:
```diff
@@ -19,8 +19,14 @@ class JwtAuthenticationFilter(
         filterChain: FilterChain
     ) {
-        val token = jwtProvider.resolveToken(request)
+        val uri = request.requestURI
+        if (uri.startsWith("/auth/kakao")) {
+            filterChain.doFilter(request, response)
+            return
+        }
```
여기도 todo 주석으로 "config에서 처리" 하나만 붙여주세요
제가 나중에 처리할게요

djlim00 replied:
SecurityConfig에서 처리해야하는데, 급하게 돌려보고 싶어서 일단 이렇게 했네요 허허 반영하겠습니다

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,44 @@
+
+    @Operation(
+        summary = "온보딩 정보 등록", description = "온보딩 화면에서 추가적으로 아르바이트 경력과 관심 직무를 회원정보에 저장합니다.\n" +
+                "관심직무는 아직 카테고리가 완전히 구현되지 않아서 추후에 추가될 예정입니다."
```
```suggestion
                "관심직무는 아직 카테고리가 완전히 구현되지 않아서 추후에 추가될 예정입니다."
```

구현되지 않았나요??

djlim00 replied:
구현됐습니다! 위 문장은 삭제하겠습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/controller/UserController.kt:
```diff
@@ -0,0 +1,44 @@
+        summary = "온보딩 정보 등록", description = "온보딩 화면에서 추가적으로 아르바이트 경력과 관심 직무를 회원정보에 저장합니다.\n" +
+                "관심직무는 아직 카테고리가 완전히 구현되지 않아서 추후에 추가될 예정입니다."
+    )
+    @PostMapping("/onboarding")
+    fun onboarding(
+        @RequestBody request: OnboardingRequest
+    ): OnboardingResponse {
+        val userId = userService.updateOnboarding(request)
+        return OnboardingResponse(userId)
+    }
+
+    @Operation(summary = "닉네임 중복확인", description = "닉네임이 이미 존재하는지 확인합니다.")
+    @GetMapping("/nickname/check")
+    fun checkNickNameDuplicate(
+        @RequestParam nickname: String
+    ): CheckNickNameResponse {
+        return userService.checkNickNameDuplicate(nickname);
+    }
```
어디서는 controller에서 userId를 받아서 객체 만들어서 반환하고
어디서는 service에서 객체를 만들어서 반환하고 통일성이 없는 상황이네요

통일시켜주시면 좋을 것 같습니다.

djlim00 replied:
전부 Service단에서 객체를 만들어서 반환하도록 반영하겠습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/entity/UserJobInterest.kt:
```diff
@@ -0,0 +1,26 @@
+    @JoinColumn(name = "user_id")
+    val user: User,
+
+    @Enumerated(EnumType.ORDINAL)
```
이거 꼭 필요한 annotation인가요??

djlim00 replied:
제가 알기로 보통 `STRING`으로 쓰는 것이 enum에 값이 추가되었을 때 문제가 생기는 것을 방지할 수 있는 것으로 알고 있습니다.
그래서 기본적으로 `ORDINAL`로 작동되긴 하지만, 명시적으로 어노테이션을 적어서 추후에 헷갈리지 않도록 적어두었습니다.

Ji-InPark replied:
기본값으로 동작하는 상황이면 전 굳이 적을 필요 없다고 생각하는 편이긴 합니다.

이건 개인 취향 차이가 있으니 판단에 맡길게요~ 반영 안하실거라면 댓글로 남겨주세요~

djlim00 replied:
저는 남겨두겠습니다!

Ji-InPark commented on src/main/resources/application.yml:
```diff
@@ -24,8 +23,8 @@ spring:
       secret-key: ${JWT_SECRET_KEY}
-      access-token-expiration-ms: 30000 # 만료시간 1시간
-      refresh-token-expiration-ms: 1209600000 # 만료시간 2주
+      access-token-expiration-ms: 300000 # 50분
```
```suggestion
      access-token-expiration-ms: 30000 # 30초
```

djlim00 committed:
chore: 엑세스 토큰 만료시간 30초로 수정

- 프론트에서 엑세스 토큰 테스트를 하기 위해 30초로 만료시간 줄였습니다.

djlim00 committed:
refactor: Auth 도매안 리팩토링

- 모든 Controller에서 Service의 함수를 return 해서 통일성을 지키도록 했습니다.
- 기존의 두가지 dto를 가지던 즉시 로그인과, 회원가입 유도가 하나의 dto로 통합되도록 했습니다.
- 리프레시 토큰을 이용해서 엑세스 토큰을 재발급 받는 dto의 이름이 모호함이 있어서 접두사를 Refresh에서 Reissue로 변경했습니다.

djlim00 committed:
chore: JwtAuthenticationFilter TODO 주석

djlim00 committed:
refactor: User 도메인 리팩토링

- UserController에서 전부 Service단의 함수를 반환하도록 했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/AuthController.kt:
```diff
@@ -0,0 +1,50 @@
+    private val authService: AuthService,
+) {
+    @Operation(
+        summary = "카카오 로그인", description = "카카오 AccessToken을 받아와 기존 회원의 경우 로그인 완료, 기존 회원이 아닌 경우 " +
+                "카카오 계정 정보를 DB에 저장만 합니다."
```
```suggestion
        summary = "카카오 로그인 및 회원가입", description = "카카오 AccessToken을 받아와 로그인을 진행합니다. 기존 회원이 아닌 경우 user 테이블에 카카오 계정 정보를 저장합니다."
```

정도로 작성해도 괜찮을 것 같아요.

djlim00 replied:
굳뜨인거 같아요. 반영하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/AuthController.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    // TODO - jwt 처리 필요
+    @Operation(summary = "회원가입", description = "회원 가입이 완료되는 시점에 카카오 정보 기반으로 사용자를 DB에 저장합니다.")
```
```suggestion
    @Operation(summary = "회원가입 정보 수정", description = "회원 가입 정보를 업데이트합니다. user 객체가 생성되지는 않습니다.")
```

회원가입에 대해서 이렇게 작성하는게 좀 더 혼선이 적지 않을까 싶네요

djlim00 replied:
사용자 입장에서는 회원가입 화면이 끝나야 회원가입하는 됐다고 생각하겠지만,, 사실 DB에는 웬만한 정보가 다 들어갔다는.. 
아무튼 개발하는 사람 입장에서는 위와 같이 수정하는게 나을 것 같네요! 반영하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/RegisterRequest.kt:
```diff
@@ -0,0 +1,21 @@
+import com.fasterxml.jackson.annotation.JsonProperty
+
+data class RegisterRequest(
+    @JsonProperty("userId")
+    val userId: Long,
+
+    @JsonProperty("nickname")
+    val nickname: String,
+
+    @JsonProperty("ageRange")
+    val ageRange: String,
+
+    @JsonProperty("gender")
+    val gender: String,
+
+    @JsonProperty("locationId")
+    val locationId: Long
```
참고로 jsonProperty와 변수명이 똑같으면 알아서 매핑해줍니다.
해당 annotation들은 필요없습니다~

djlim00 replied:
다른 dto들과 다르게 여기만 이걸 넣은 이유가. 처음에 다른 dto들과 동일하게 `@JsonProperty("")`을 안적고 하니까
`Cannot construct instance of com.arabyte.arabyteapi.domain.auth.dto.RegisterRequest (no Creators, like default constructor, exist)` 
위의 에러가 계속 발생하더라구요. 
코틀린은 자바랑 다르게 런타임에 기본생성자를 만들어주지 않아서, Jackson이 파라미터를 하나라도 못 받았을 때 기본값을 채우면서 생성할 수가 없어서 생기는 오류라고 하는데, 이걸 해결하는 방법이 `@JsonProperty("")` 어노테이션을 달아주면 된다고 합니다.

djlim00 replied:
확인해본 결과 지금 모든 Request DTO에서 `@JsonProperty("")`를 안붙이면 제대로 동작하지 않고 있습니다.

djlim00 replied:
`data class CreateArticleRequest @JsonCreator constructor` 이런식으로 클래스 단위에 @JsonCreator를 달면 생성자가 생겨서 모든 파라미터에 `@JsonProperty`를 달지 않아도 제대로 작동되는데, Jackson에서 왜 인식을 못하는건지 모르겠네요

djlim00 replied:
https://traeper.tistory.com/209
위 링크 참고하여, build.gradle.kts에 `implementation("com.fasterxml.jackson.module:jackson-module-kotlin")` 추가하고 
JacksonConfiguration에 `registerKotlinModule()` 추가여서 해결했습니다.

Ji-InPark replied:
키오스쿨에서도 implementation("com.fasterxml.jackson.module:jackson-module-kotlin")는 사용중인데 
registerKotlinModule() 은 사용하지 않고 있어서요.
혹시 registerKotlinModule() 빼도 잘 돌아가나요??

djlim00 replied:
registerKotlinModule() 빼면 안돌아가더라구요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/AuthService.kt:
```diff
@@ -0,0 +1,95 @@
+                refreshToken = jwtProvider.generateRefreshToken(user.id.toString()),
+                isRegistered = false
+            )
+        } else {
+            return AuthorizeResponse(
+                userId = user.id,
+                accessToken = jwtProvider.generateAccessToken(user.id.toString()),
+                refreshToken = jwtProvider.generateRefreshToken(user.id.toString()),
+                isRegistered = true
+            )
+        }
```
개인적으로는 else를 빼는걸 좋아하긴 합니다.

djlim00 replied:
좀 더 깔끔하네용 반영할게요~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/controller/AuthController.kt:
```diff
@@ -0,0 +1,50 @@
+    @Operation(summary = "액세스 토큰 갱신", description = "refresh token을 이용하여 access token을 갱신합니다.")
+    @PostMapping("/refresh")
+    @SwaggerCustomException(CustomExceptionGroup.REFRESH_ACCESS_TOKEN)
+    fun refreshAccessToken(
+        @RequestBody request: ReissueAccessTokenRequest
+    ): ReissueAccessTokenResponse {
```
request하고 response 이름을 함수명과 다르게 가져간 이유가 있나요?

djlim00 replied:
shit 함수명을 안바꾼겁니다. 감사합니다. 반영하겠습니다!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/dto/OnboardingRequest.kt:
```diff
@@ -0,0 +1,25 @@
+    @JsonProperty("experience_months")
+    val experienceMonths: Int,
+
+    @JsonProperty("job_interests_1")
+    val jobInterests1: Category,
+
+    @JsonProperty("job_interests_2")
+    val jobInterests2: Category,
+
+    @JsonProperty("job_interests_3")
+    val jobInterests3: Category
```
이렇게 받으면 직업을 3개 이하로 받았을 때 문제가 생길것 같아요.
그리고 안드로이드에서도 parameter 보낼 때 조금 귀찮을 것 같습니다. (list에 담아두었던것을 따로 빼내서 1, 2, 3에 각각 할당해야하기 때문이죠)
우리쪽에서 list로 받아서 처리하는게 더 좋을 것 같습니다.

djlim00 replied:
넵 List로 변경해서 받도록 수정하겠습니다.
또 온보딩을 건너뛸수도 있고, 경력만 입력하고 관심직무를 입력하지 않는 경우가 있어서 null처리도 좀 하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/user/service/UserService.kt:
```diff
@@ -18,8 +24,52 @@ class UserService(
+
+    fun checkNickNameDuplicate(nickname: String): CheckNickNameResponse {
+        val isDuplicate = isNicknameExists(nickname)
+        val message = if (isDuplicate) "닉네임이 중복됩니다." else "사용 가능한 닉네임입니다."
```
전 이정도 메시지는 프론트에 맡기는 편이긴 합니다.

djlim00 replied:
글킨해요.  isDuplicate만 반환하도록 하겠습니다.

Ji-InPark reviewed (commented):
거의 다 왔습니다!!

djlim00 committed:
bugfix: Jackson Json 역직렬화 문제 해결

- jackson-module-kotlin을 추가하여 Json역직렬화가 안되는 문제를 해결했습니다.
- 기존의 에러를 해결하기 위해 걸어두었던 어노테이션을 삭제했습니다.

djlim00 committed:
refactor: Auth관련 리팩토링

- 스웨거 설명을 수정했습니다.
- 함수명과 이름이 달라서 통일성 없는Controller의 함수명과 url을 수정했습니다.
- 필요없는 DTO 삭제했습니다.

djlim00 committed:
update: User관련 코드 업데이트

- 온보딩 관련 로직 중 온보딩을 건너뛰는 경우 null 처리와 데이터베이스 nullable 여부 변경, User 엔티티 null 가능 여부 수정
- 닉네임 중복 반환값을 boolean 값만 반환하도록 수정했습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/service/AuthService.kt:
```diff
@@ -0,0 +1,95 @@
+        )
+    }
+
+    fun getUserIdByToken(body: ReissueAccessTokenRequest): String {
```
```suggestion
    private fun getUserIdByToken(body: ReissueAccessTokenRequest): String {
```

djlim00 replied:
shit 발견 못했네요 반영하고 머지 할게요

Ji-InPark reviewed (approved):
![image](https://github.com/user-attachments/assets/259d3c4d-a78c-47c8-982d-4aaf423dc585)

반영하고 머지해주세요! 수고하셨습니다!!

djlim00 committed:
Update AuthService.kt
